### PR TITLE
Optimize the decision dag by eliminating semantically identical states.

### PR DESCRIPTION
**Pull request by gafter**
* Optimize the decision dag by eliminating semantically identical states.
* Change the order of nodes while computing the state machine to match the order in the final code
* Implement `is not`.
* Stop using `operator ==` and use `.Equals` instead.
* Remaining work tracked at https://github.com/dotnet/roslyn/issues/41502

### CODE REVIEW

**Comment by gafter** [comment]
@agocke @RikkiGibson Could you please review this pattern-matching change?

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagTest.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using System.Diagnostics.CodeAnalysis;
+using Roslyn.Utilities;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundDagTest
+    {
+        public override bool Equals([NotNullWhen(true)] object? obj) => obj is BoundDagTest other && this.Equals(other);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagTest.cs]: it's expected that we would learn that 'obj' is not null when an override of `object.Equals(object)` returns true, even without the attribute. Is that not the case now?

File: src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDagNode.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.CodeAnalysis;
+using Roslyn.Utilities;
+
+#nullable enable
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDagNode.cs]: the convention adopted in much of the nullable-annotated code already in Roslyn is to put the #nullable directive after the license and before any usings. e.g. [https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Symbols/MethodKind.cs](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Symbols/MethodKind.cs)

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagTest.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundDagTest.cs]: I have no idea.  This was essentially copied from elsewhere in the compiler.  If we eliminate this attribute for this use, I would expect we'd do that here and elsewhere.

---
In reply to: [385934220](https://github.com/dotnet/roslyn/pull/42012#discussion_r385934220) [](ancestors = 385934220)

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs [hunk 3]
```diff
@@ -1177,7 +1227,7 @@ private static ImmutableArray<DagState> Successor(DagState state)
 
                 if (state.TrueBranch != null && state.FalseBranch != null)
                 {
-                    return ImmutableArray.Create(state.TrueBranch, state.FalseBranch);
+                    return ImmutableArray.Create(state.FalseBranch, state.TrueBranch);
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs]: We want to produce the successors in the same order that we generate the code in the end.  The two successor implementations currently disagree.

---
In reply to: [385956147](https://github.com/dotnet/roslyn/pull/42012#discussion_r385956147) [](ancestors = 385956147)

File: src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDagNode.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                case (BoundLeafDecisionDagNode n1, BoundLeafDecisionDagNode n2):
+                    return n1.Label == n2.Label;
+                default:
+                    return false;
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            switch (this)
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDagNode.cs]: Too hard to debug.

---
In reply to: [385957790](https://github.com/dotnet/roslyn/pull/42012#discussion_r385957790) [](ancestors = 385957790)

**Comment by gafter** [comment]
@agocke @RikkiGibson I think I've responded to all of your comments.  Do you have any others?

**Comment by gafter** [comment]
@agocke @RikkiGibson I think I've responded to all of your comments. Do you have any others?

File: src/Compilers/Core/Portable/Collections/DictionaryExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            if (dictionary.TryGetValue(key, out var existingValue))
+            {
+                return existingValue;
+            }
+            else
+            {
+                dictionary.Add(key, value);
+                return value;
+            }
```
Comment (agocke) [inline src/Compilers/Core/Portable/Collections/DictionaryExtensions.cs]:
```suggestion
            if (!dictionary.TryGetValue(key, out var existingValue))
            {
                existingValue = value;
                dictionary.Add(key, existingValue);
            }
            return existingValue;
```

Comment (gafter) [reply src/Compilers/Core/Portable/Collections/DictionaryExtensions.cs]: It took me a few minutes to figure out what your code was doing. I find it counterintuitive to use a pattern-variable as a mutable local where it isn't definitely assigned.  Also, the name doesn't really make sense that way.  I'd rather leave it as it is.

---
In reply to: [387411596](https://github.com/dotnet/roslyn/pull/42012#discussion_r387411596) [](ancestors = 387411596)

Comment (agocke) [reply src/Compilers/Core/Portable/Collections/DictionaryExtensions.cs]: Sounds good, not wed to it

Comment (agocke) [reply src/Compilers/Core/Portable/Collections/DictionaryExtensions.cs]: Alternatively, you could use `out value` to remove the pattern variable

**Pull request merged**